drivers/i2c/i2c_bitbang.c

```diff
@@ -0,0 +1,340 @@
+static void i2c_bitbang_set_sda(FAR struct i2c_bitbang_dev_s *dev,
+                                bool high);
+
+static int i2c_bitbang_wait_ack(FAR struct i2c_bitbang_dev_s *dev);
+static void i2c_bitbang_send(FAR struct i2c_bitbang_dev_s *dev,
+                             uint8_t data);
+
+/****************************************************************************
+ * Private Data
+ ****************************************************************************/
+
+struct i2c_ops_s g_i2c_ops =
+{
+  .transfer = i2c_bitbang_transfer
+};
+
+/****************************************************************************
+ * Private Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: i2c_bitbang_transfer
+ ****************************************************************************/
+
+static int i2c_bitbang_transfer(FAR struct i2c_master_s *dev,
+                                FAR struct i2c_msg_s *msgs, int count)
+{
+  FAR struct i2c_bitbang_dev_s *priv = (FAR struct i2c_bitbang_dev_s *)dev;
+  int ret = OK;
+  int i;
+
+  for (i = 0; i < count; i++)
+    {
+      uint8_t addr;
+      FAR struct i2c_msg_s *msg = &msgs[i];
+
+#ifdef CONFIG_I2C_BITBANG_NO_DELAY
+      /* Compute delay from frequency */
+
+      priv->delay = ((1000000 / msg->frequency) / 2) -
+                    CONFIG_I2C_BITBANG_GPIO_OVERHEAD;
+
+      if (priv->delay < 0)
+        {
+          priv->delay = 0;
+        }
+#endif
+
+      /* Send start bit */
+
+      i2c_bitbang_set_scl(priv, true);
+      i2c_bitbang_set_sda(priv, false);
+      i2c_bitbang_set_scl(priv, false);
+
+      /* Send the address */
+
+      addr = (msg->flags & I2C_M_READ ? I2C_READADDR8(msg->addr) :
+                                        I2C_WRITEADDR8(msg->addr));
+
+      i2c_bitbang_send(priv, addr);
+
+      /* Wait for ACK */
+
+      ret = i2c_bitbang_wait_ack(priv);
+
+      if (ret < 0)
+        {
+          return ret;
+        }
+
+      i2c_bitbang_set_scl(priv, false);
+
+      if (msgs->flags & I2C_M_READ)
+        {
+          int j;
+
+          for (j = 0; j < msg->length; j++)
+            {
+              /* Send the data */
+
+              i2c_bitbang_send(priv, msg->buffer[j]);
+
+              ret = i2c_bitbang_wait_ack(priv);
+
+              if (ret < 0)
+                {
+                  return ret;
+                }
+
+              i2c_bitbang_set_scl(priv, false);
+            }
+        }
+      else
+        {
+          int j;
+          int k;
+
+          for (j = 0; j < msg->length; j++)
+            {
+              uint8_t data = 0;
+
+              i2c_bitbang_set_sda(priv, true);
+
+              msg->buffer[j] = 0;
+
+              for (k = 0; k < 8; k++)
+                {
+                  i2c_bitbang_set_scl(priv, true);
+                  data |= priv->lower->ops->get_sda(priv->lower) << (7 - k);
```

Comment:
add a wrapper function for get_sda?

Reply:
It would only be a direct call. I can make a define or inline to avoid extra overhead, but it would be mostly cosmetic.